doc/img2txt.1
.TH img2txt 1 "2007-11-07" "libcaca"
.SH NAME
img2txt \- convert images to various text-based coloured files
.SH SYNOPSIS
.B img2txt 
[
.B \-W
.I width
]
[
.B \-H
.I height
]
[
.B \-x
.I font-width
]
[
.B \-y
.I font-height
]
.PD 0
.IP
.PD
 [
.B \-b
.I brightness
]
[
.B \-c
.I contrast
]
[
.B \-g
.I gamma
]
.PD 0
.IP
.PD
 [
.B \-d
.I dither
]
[
.B \-f
.I format
] 
.I FILE
.RI
.SH DESCRIPTION
.B img2txt
converts images to colour ASCII characters and outputs them to text-based
coloured files.
.PP
.B img2txt
can load the most widespread image formats: PNG, JPEG, GIF, PNG, BMP etc (see 
.B NOTES
for details).
By default the output text is 60 columns wide, and the line count is 
computed accordingly to respect aspect ratio of original file. 
The default output format is standard ANSI coloured text.

.SH OPTIONS
.TP
.B \-W, \-\-width=<width>
Change output column count. If not given, the default is set to 60.
.TP
.B \-H, \-\-height=<height>
Change output line count. If not given, the height is computed to match correct aspect ratio.
.TP
.B \-x, \-\-font-width=<width>
Change output font width. If not given, the default is set to 6. This value will be used for computing aspect ratio.
.TP
.B \-y, \-\-font-height=<height>
Change output font height. If not given, the default is set to 10. This value will be used for computing aspect ratio.
.TP
.B \-b, \-\-brightness=<brightness>
Change image brightness. Default to 1.0.
.TP
.B \-c, \-\-contrast=<contrast>
Change image contrast. Default to 1.0.
.TP
.B \-g, \-\-gamma=<gamma>
Change image gamma. Default to 1.0.
.TP
.B \-d, \-\-dither=<dither>
Change dithering algorithm. This can be one of the following (default to fstein) :
  none     : Nearest color
  ordered2 : Ordered 2x2
  ordered4 : Ordered 4x4
  ordered8 : Ordered 8x8
  random   : Random
  fstein   : Floyd Steinberg
.TP
.B \-f, \-\-format=<format>
Change output format. This can be one of the following (default to ansi) :
  ansi   : coloured ANSI 
  caca   : internal libcaca format
  utf8   : UTF8 with CR
  utf8cr : UTF8 with CRLF (MS Windows)
  html   : HTML with CSS and DIV support
  html3  : Pure HTML3 with tables
  irc    : IRC with ctrl-k codes
  bbfr   : BBCode (French)
  ps     : Postscript
  svg    : Scalable Vector Graphics
  tga    : Targa Image
.TP
.B \-h, \-\-help
Display help message and exit.
.TP
.B \-v, \-\-version
Display version of the program and exit.

.SH EXAMPLES
img2txt hello.jpg > hello.ans

img2txt \-\-width=40 \-\-format=svg hello.jpg > tinyhello.svg

.SH NOTES
Setting both column and line count (using 
\-\-width
and
\-\-height) will let you choose the exact output size without taking aspect ratio in account.

You must compile libcaca package with support of
.I Imlib2
to be able to load a wide variety of image formats. Otherwise you will only  be able to load regular BMP files.

.SH SEE ALSO
cacaview(1)
.SH AUTHOR
This manual page was written by Sam Hocevar <sam@hocevar.net> and Jean-Yves Lamoureux <jylam@lnxscene.org>.


NEWS
/** \page libcaca-news Libcaca news

\section news0_99_beta18 Changes between 0.99.beta18 and 0.99.beta17

 - UNRELEASED

\section news0_99_beta17 Changes between 0.99.beta17 and 0.99.beta16

 - new dirty rectangle framework for accelerated rendering
 - new <conio.h> compatible header for old DOS software porting
 - XCode and Visual Studio build improvements
 - Unicode support in the X11 driver
 - triangle texture mapping
 - improved HTML output
 - new troff output
 - improved Ruby and .NET bindings
 - new comprehensive PHP bindings
 - new Java bindings
 - Cocoa driver enhancements
 - support for ANSI colour in FIGlet and TOIlet fonts
 - documentation updates
 - numerous improvements and bugfixes

\section news0_99_beta16 Changes between 0.99.beta16 and 0.99.beta15

 - many build fixes, especially for nonstandard platforms

\section news0_99_beta15 Changes between 0.99.beta15 and 0.99.beta14

 - libcucul was merged back into libcaca for more clarity

\section news0_99_beta14 Changes between 0.99.beta14 and 0.99.beta13

 - internal FIGlet font support
 - use C99 types in public headers
 - runtime support for output drivers
 - BBcode export support

\section news0_99_beta13 Changes between 0.99.beta13 and 0.99.beta12

 - device-dependent cursor support
 - event API rewrite
 - minor API improvements and extensions
 - img2txt improvements
 - Ruby bindings
 - Massive C# bindings improvements
 - Python sample code
 - Visual Studio build solution

\section news0_99_beta12 Changes between 0.99.beta12 and 0.99.beta11

 - support for 90-degree canvas rotation
 - better behaviour when trying to output Unicode on an ASCII terminal
 - the built-in font now supports the Geometric Shapes, Halfwidth and Fullwidth
   Forms and Miscellaneous Symbols Unicode blocks
 - new rotozoom effect in cacademo
 - Cocoa output driver for Mac OS X
 - preliminary .NET bindings
 - many bugfixes and documentation changes

\section news0_99_beta11 Changes between 0.99.beta11 and 0.99.beta10

 - fixed compilation of the C++ bindings
 - fixed bugs in cucul_import_memory(), cucul_set_canvas_size()
 - implemented caca_set_display_title() for ncurses and S-Lang
 - minor bugfixes

\section news0_99_beta10 Changes between 0.99.beta10 and 0.99.beta9

 - new debug mode
 - blitting canvases now makes use of the canvas' handle coordinates
 - import functions can read streamed data
 - attribute to colorspace transformations
 - added katakana and hiragana glyphs to the built-in font
 - many bugfixes and documentation changes

\section news0_99_beta9 Changes between 0.99.beta9 and 0.99.beta8

 - support for blink, bold, italics and underline attributes
 - allow to import and export zero-sized canvases
 - fixed Imlib2 support in cacaview
 - fixed buffer overlows in the file importer
 - big documentation updates

\section news0_99_beta8 Changes between 0.99.beta8 and 0.99.beta7

 - allow to build the X11 and GL drivers as separate plugins
 - support for fullwidth Unicode characters
 - improved cucul_flip() and cucul_rotate()
 - minor bugfixes and documentation updates

\section news0_99_beta7 Changes between 0.99.beta7 and 0.99.beta6

 - transparency support in the UTF-8 importer and exporter
 - optimised triangle fill routine
 - updated C++ bindings

\section news0_99_beta6 Changes between 0.99.beta6 and 0.99.beta5

 - ANSI importer now handles transparency and UTF-8
 - Unicode support was broken for about 10% of the set
 - various memory leaks were fixed

\section news0_99_beta5 Changes between 0.99.beta5 and 0.99.beta4

 - implemented cucul_getchar() and cucul_get_color()
 - handle transparency in the IRC export
 - new cropping and expanding filters
 - full Unicode support in the OpenGL driver
 - portability fixes for 64-bit architectures, Win32 and MS-DOS
 - all demos except cacafire were merged into cacademo

\section news0_99_beta4 Changes between 0.99.beta4 and 0.99.beta3

 - added a compatibility layer for pre-1.x libcaca applications
 - fixed manpage generation
 - minor bugfixes and documentation updates

\section news0_99_beta3 Changes between 0.99.beta3 and 0.99.beta2

 - libcaca functions use errno for error reporting
 - updated C++ bindings
 - minor improvements, bugfixes and documentation updates

\section news0_99_beta2 Changes between 0.99.beta2 and 0.99.beta1

 - ANSI importer
 - functions use errno for error reporting
 - updated C++ bindings
 - .NET bindings
 - cacadraw, an ANSI viewer that will evolve into an editor
 - Unicode input and output support for SLang and ncurses
 - built-in fonts work on Win32

\section news0_99_beta1 Changes between 0.9 and 0.99.beta1

 - license switched to WTFPL
 - libcaca was split into libcucul, a standalone text manipulation backend,
    and libcaca, the display and user input frontend
 - Unicode support
 - TrueColor (more than 16 colours) support
 - Floyd-Steinberg dithering
 - gamma correction
 - export functions for HTML, IRC, ANSI, SVG, PostScript, TGA...
 - builtin fonts for device-independent bitmap output
 - various text transformation routines (rotation, horizontal flip...)
 - OpenGL renderer
 - kernel mode to build libcaca programs into a bootable x86 kernel
 - cacaserver, a telnet server that can be hooked to libcaca applications
 - img2irc, an image to IRC conversion utility

\section news0_9 Changes between 0.8 and 0.9

 - fix for a buffer overflow in the line rendering
 - fixed resizing in the ncurses and slang drivers
 - aspect ratio and finer zoom support in cacaview
 - minor compilation fixes

\section news0_8 Changes between 0.7 and 0.8

 - window resizing support
 - native Win32 port
 - autorepeat emulation in the ncurses and slang drivers
 - support for more keycodes in the ncurses and slang drivers
 - cacaplas, a plasma animation example
 - cacamoir, a moire circles animation example
 - MSVC project file

\section news0_7 Changes between 0.6 and 0.7

 - many bugfixes in the event handling
 - cacaball, a metaball animation example


src/img2txt.c
/*
 *  img2txt       image to text converter
 *  Copyright (c) 2006-2010 Sam Hocevar <sam@hocevar.net>
 *                2007 Jean-Yves Lamoureux <jylam@lnxscene.org>
 *                All Rights Reserved
 *
 *  This program is free software. It comes without any warranty, to
 *  the extent permitted by applicable law. You can redistribute it
 *  and/or modify it under the terms of the Do What The Fuck You Want
 *  To Public License, Version 2, as published by Sam Hocevar. See
 *  http://sam.zoy.org/wtfpl/COPYING for more details.
 */

#include "config.h"

#if !defined(__KERNEL__)
#   include <stdio.h>
#   include <string.h>
#   include <stdlib.h>
#endif

#include "caca.h"

#include "common-image.h"

#define IMG2TXTVERSION "0.1"

static void usage(int argc, char **argv)
{
    char const * const * list;

    fprintf(stderr, "Usage: %s [OPTIONS]... <IMAGE>\n", argv[0]);
    fprintf(stderr, "Convert IMAGE to any text based available format.\n");
    fprintf(stderr, "Example : %s -w 80 -f ansi ./caca.png\n\n", argv[0]);
    fprintf(stderr, "Options:\n");
    fprintf(stderr, "  -h, --help\t\t\tThis help\n");
    fprintf(stderr, "  -v, --version\t\t\tVersion of the program\n");
    fprintf(stderr, "  -W, --width=WIDTH\t\tWidth of resulting image\n");
    fprintf(stderr, "  -H, --height=HEIGHT\t\tHeight of resulting image\n");
    fprintf(stderr, "  -x, --font-width=WIDTH\t\tWidth of output font\n");
    fprintf(stderr, "  -y, --font-height=HEIGHT\t\tHeight of output font\n");
    fprintf(stderr, "  -b, --brightness=BRIGHTNESS\tBrightness of resulting image\n");
    fprintf(stderr, "  -c, --contrast=CONTRAST\tContrast of resulting image\n");
    fprintf(stderr, "  -g, --gamma=GAMMA\t\tGamma of resulting image\n");
    fprintf(stderr, "  -d, --dither=DITHER\t\tDithering algorithm to use :\n");
    list = caca_get_dither_algorithm_list(NULL);
    while(*list)
    {
        fprintf(stderr, "\t\t\t%s: %s\n", list[0], list[1]);
        list += 2;
    }

    fprintf(stderr, "  -f, --format=FORMAT\t\tFormat of the resulting image :\n");
    list = caca_get_export_list();
    while(*list)
    {
        fprintf(stderr, "\t\t\t%s: %s\n", list[0], list[1]);
        list += 2;
    }

#if !defined(USE_IMLIB2)
    fprintf(stderr, "NOTE: This program has NOT been built with Imlib2 support. Only BMP loading is supported.\n");
#endif
}

static void version(void)
{
    printf(
    "img2txt Copyright 2006-2007 Sam Hocevar and Jean-Yves Lamoureux\n"
    "Internet: <sam@hocevar.net> <jylam@lnxscene.org> Version: %s, date: %s\n"
    "\n"
    "img2txt, along with its documentation, may be freely copied and distributed.\n"
    "\n"
    "The latest version of img2txt is available from the web site,\n"
    "        http://caca.zoy.org/wiki/libcaca in the libcaca package.\n"
    "\n",
    caca_get_version(), __DATE__);
}
int main(int argc, char **argv)
{
    /* libcaca context */
    caca_canvas_t *cv;
    void *export;
    size_t len;
    struct image *i;
    unsigned int cols = 0, lines = 0, font_width = 6, font_height = 10;
    char *format = NULL;
    char *dither = NULL;
    float gamma = -1, brightness = -1, contrast = -1;

    if(argc < 2)
    {
        fprintf(stderr, "%s: wrong argument count\n", argv[0]);
        usage(argc, argv);
        return 1;
    }

    for(;;)
    {
        int option_index = 0;
        static struct caca_option long_options[] =
        {
            { "width",       1, NULL, 'W' },
            { "height",      1, NULL, 'H' },
            { "font-width",  1, NULL, 'x' },
            { "font-height", 1, NULL, 'y' },
            { "format",      1, NULL, 'f' },
            { "dither",      1, NULL, 'd' },
            { "gamma",       1, NULL, 'g' },
            { "brightness",  1, NULL, 'b' },
            { "contrast",    1, NULL, 'c' },
            { "help",        0, NULL, 'h' },
            { "version",     0, NULL, 'v' },
        };
        int c = caca_getopt(argc, argv, "W:H:f:d:g:b:c:hvx:y:",
                            long_options, &option_index);
        if(c == -1)
            break;

        switch(c)
        {
        case 'W': /* --width */
            cols = atoi(caca_optarg);
            break;
        case 'H': /* --height */
            lines = atoi(caca_optarg);
            break;
        case 'x': /* --width */
            font_width = atoi(caca_optarg);
            break;
        case 'y': /* --height */
            font_height = atoi(caca_optarg);
            break;
        case 'f': /* --format */
            format = caca_optarg;
            break;
        case 'd': /* --dither */
            dither = caca_optarg;
            break;
        case 'g': /* --gamma */
            gamma = atof(caca_optarg);
            break;
        case 'b': /* --brightness */
            brightness = atof(caca_optarg);
            break;
        case 'c': /* --contrast */
            contrast = atof(caca_optarg);
            break;
        case 'h': /* --help */
            usage(argc, argv);
            return 0;
            break;
        case 'v': /* --version */
            version();
            return 0;
            break;
        default:
            return 1;
            break;
        }
    }


    cv = caca_create_canvas(0, 0);
    if(!cv)
    {
        fprintf(stderr, "%s: unable to initialise libcaca\n", argv[0]);
        return 1;
    }

    i = load_image(argv[argc-1]);
    if(!i)
    {
        fprintf(stderr, "%s: unable to load %s\n", argv[0], argv[argc-1]);
        caca_free_canvas(cv);
        return 1;
    }

    /* Assume a 6×10 font */
    if(!cols && !lines)
    {
        cols = 60;
        lines = cols * i->h * font_width / i->w / font_height;
    }
    else if(cols && !lines)
    {
        lines = cols * i->h * font_width / i->w / font_height;
    }
    else if(!cols && lines)
    {
        cols = lines * i->w * font_height / i->h / font_width;
    }


    caca_set_canvas_size(cv, cols, lines);
    caca_set_color_ansi(cv, CACA_DEFAULT, CACA_TRANSPARENT);
    caca_clear_canvas(cv);
    if(caca_set_dither_algorithm(i->dither, dither?dither:"fstein"))
    {
        fprintf(stderr, "%s: Can't dither image with algorithm '%s'\n", argv[0], dither);
        unload_image(i);
        caca_free_canvas(cv);
        return -1;
    }

    if(brightness!=-1) caca_set_dither_brightness (i->dither, brightness);
    if(contrast!=-1) caca_set_dither_contrast (i->dither, contrast);
    if(gamma!=-1) caca_set_dither_gamma (i->dither, gamma);

    caca_dither_bitmap(cv, 0, 0, cols, lines, i->dither, i->pixels);

    unload_image(i);

    export = caca_export_canvas_to_memory(cv, format?format:"ansi", &len);
    if(!export)
    {
        fprintf(stderr, "%s: Can't export to format '%s'\n", argv[0], format);
    }
    else
    {
        fwrite(export, len, 1, stdout);
        free(export);
    }

    caca_free_canvas(cv);

    return 0;
}

